model/riscv_vmem.sail

```diff
@@ -0,0 +1,851 @@
+       // SATP CSR
+       satp_asid_size_bits = 16,
+       satp_asid_lsb_index = 44,
+
+       // SATP PPN
+       satp_ppn_size_bits  = 44,
+       satp_ppn_lsb_index  = 0,
+
+       // VA
+       va_size_bits	   = 39,
+       vpn_size_bits	   = 9,
+
+       // PTE
+       levels		   = 3,
+       log_pte_size_bytes  = 3,	   // 8 Bytes
+       pte_msbs_lsb_index  = 54,
+       pte_msbs_size_bits  = 10,
+       pte_PPNs_lsb_index  = 10,
+       pte_PPNs_size_bits  = 44,
+       pte_PPN_j_size_bits = 9
+}
+
+// PRIVATE
+let sv48_params : SV_Params = struct {
+       // SATP CSR
+       satp_asid_size_bits = 16,
+       satp_asid_lsb_index = 44,
+
+       // SATP PPN
+       satp_ppn_size_bits  = 44,
+       satp_ppn_lsb_index  = 0,
+
+       // VA
+       va_size_bits	   = 48,
+       vpn_size_bits	   = 9,
+
+       // PTE
+       levels		   = 4,
+       log_pte_size_bytes  = 3,	   // 8 Bytes
+       pte_msbs_lsb_index  = 54,
+       pte_msbs_size_bits  = 10,
+       pte_PPNs_lsb_index  = 10,
+       pte_PPNs_size_bits  = 44,
+       pte_PPN_j_size_bits = 9
+}
+
+// For future; not currently used
+// PRIVATE
+/*
+let sv57_params : SV_Params = struct {
+       // SATP CSR
+       satp_asid_size_bits = 16,
+       satp_asid_lsb_index = 44,
+
+       // SATP PPN
+       satp_ppn_size_bits  = 44,
+       satp_ppn_lsb_index  = 0,
+
+       // VA
+       va_size_bits	   = 57,
+       vpn_size_bits	   = 9,
+
+       // PTE
+       levels		   = 5,
+       log_pte_size_bytes  = 3,	   // 8 Bytes
+       pte_msbs_lsb_index  = 54,
+       pte_msbs_size_bits  = 10,
+       pte_PPNs_lsb_index  = 10,
+       pte_PPNs_size_bits  = 44,
+       pte_PPN_j_size_bits = 9
+}
+*/
+
+// This 'undefined_SV_Params()' function is not used anywhere, but is
+// currently (2023-12) needed to work around an issue where Sail tries
+// to figure out how it could do
+//     let x : SV_Params = undefined
+// even though the code never does this.  This has been fixed in Sail.
+// The fix will become available in a new Sail release, at which point
+// this function can be deleted (TODO).
+// PRIVATE
+val	 undefined_SV_Params : unit -> SV_Params
+function undefined_SV_Params() = sv32_params
+
+// ****************************************************************
+// Fields of virtual addresses
+
+// PRIVATE: Extract full VPN field from VA
+function vpns_of_va(sv_params : SV_Params,
+		    va	      : bits(64)) -> bits(64) = {
+  let mask : bits(64) = zero_extend(ones(sv_params.va_size_bits));
+  (va & mask) >> PAGESIZE_BITS
+}
+
+// PRIVATE: Extract VPN[level] from  VA
+function vpn_j_of_va(sv_params : SV_Params,
+		     va	       : bits(64),
+		     level     : nat)	    -> bits(64) = {
+  let lsb  : nat       = PAGESIZE_BITS + level * sv_params.vpn_size_bits;
+  let mask : bits(64) = zero_extend(ones(sv_params.vpn_size_bits));
+  ((va >> lsb) & mask)
+}
+
+// PRIVATE: Extract offset within page from VA
+function offset_of_va(va : bits(64)) -> bits(64) = {
```

Comment:
```
function offset_of_va(va : bits(64)) -> bits(PAGESIZE_BITS) = va[PAGESIZE_BITS - 1 .. 0]
```

Reply:
Ok (shrinking output type from bits(64) to bits[PAGESIZE_BITS] needed a compensating zero_extend() in the two places where it was invoked).